aspnetcore/blazor/components.md

```diff
@@ -541,6 +541,56 @@ While capturing component references use a similar syntax to [capturing element
 > [!NOTE]
 > Do **not** use component references to mutate the state of child components. Instead, use normal declarative parameters to pass data to child components. Use of normal declarative parameters result in child components that rerender at the correct times automatically.
 
+## Invoke component methods externally to update state
+
+A component method can directly or indirectly (from another component or a service) be used to update state. Consider a *notifier service* that can notify any listening component of the updated state:
+
+```csharp
+public class NotifierService
+{
+    // Can be called from anywhere
+    public void Update(string key, int value)
+    {
+        Notify?.Invoke(key, value);
+    }
+
+    public event Action<string, int> Notify;
+}
+```
+
+Usage of the `NotifierService` to update a component:
+
+```cshtml
+@page "/"
+@inject NotifierService Notifier
+@implements IDisposable
+
+<p>Last update: @lastNotification.key = @lastNotification.value</p>
+
+@code {
+    private (string key, int value) lastNotification;
+
+    protected override void OnInitialized()
+    {
+        Notifier.Notify += OnNotify;
+    }
+
+    public async void OnNotify(string key, int value)
+    {
+        lastNotification = (key, value);
+
+        await InvokeAsync(StateHasChanged);
+    }
```

Comment:
Possible improvement:

```cs
    async Task OnNotifyAsync()
    {
        return InvokeAsync(() =>
        {
            lastNotification = (key, value);
            StateHasChanged();
        });
    }
```

This is better because:

 * It moves the error handling responsibility upstream. We're still ignoring any errors (like you had `async void` before), but now if you *wanted* to know about them from the caller you could.
 * It avoids potential thread safety issues. If instead of `lastNotification = (key, value)` you had a large block of complex logic, you might not realise there could be multiple concurrent invocations, given that most component logic assumes one-concurrent-execution within the sync context. In most cases it's better to run your update logic on the sync context to avoid needing to reason about different thread safety considerations.

If you do make this change, you'll need to change the caller because the signature is different. Maybe it's also possible to have the caller handle exceptions in some sensible way but I'm not sure what (just log them?).

Reply:
Not sure what you mean by "a lot of complexity". I'd suggest updating the code to match what I posted, but not necessarily also adding all the explanations I've given (I only posted that as justification for us).

Even if we're not explaining all the reasons for why our code samples are as they are, it's still good for them to be robust where possible so people can safely copy/paste them, right?

Reply:
Oh perhaps you were referring to the thoughts about error handling. I don't think it's necessary to put that in these samples really. We'd just need to change the calling code to handle the fact that it's invoking a method called `OnNotifyAsync` instead of `OnNotify`.

Reply:
I'm in favor of the changes @SteveSandersonMS is proposing here.

Reply:
Made a couple of small tweaks to the text.